### PR TITLE
Google Analytics

### PR DESCRIPTION
**Pull request by lalitlj**
Snake Case

### CODE REVIEW

File: googleanalytics/userReport.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    RunReportRequest,
+)
+
+class UserReportInput(BaseModel):
+    start: str = Field(..., description="The starting date of the query, in YYYY-MM-DD format")
+    end: str = Field(..., description="The last date of the query, in YYYY-MM-DD format, if today, return today's date")
+    issinglefile: bool= Field(..., description="True if it is specified that the report should be stored in a single file. False otherwise.")
+
+
+class reportTool(BaseTool):
```
Comment (TransformerOptimus) [inline googleanalytics/userReport.py]: rename the class to to GoogleAnalyticsReportTool and
GoogleAnalyticsReportToolInput

File: googleanalytics/userReport.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    DateRange,
+    Dimension,
+    Metric,
+    RunReportRequest,
+)
+
+class UserReportInput(BaseModel):
+    start: str = Field(..., description="The starting date of the query, in YYYY-MM-DD format")
+    end: str = Field(..., description="The last date of the query, in YYYY-MM-DD format, if today, return today's date")
+    issinglefile: bool= Field(..., description="True if it is specified that the report should be stored in a single file. False otherwise.")
```
Comment (TransformerOptimus) [inline googleanalytics/userReport.py]: is_single_file

File: googleanalytics/userReport.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+        response = ""
+        for name in listofnames:
+            response=response +" "+ name+","
+
+        response=response[:-1]+"."
+
+        return "Succesfully wrote"+response
+
+    def returnDimMetrics(self):
```
Comment (TransformerOptimus) [inline googleanalytics/userReport.py]: function should be snake case.

File: googleanalytics/userReport.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        DimMetrics = []
+        # try:
+        with open("superagi/tools/marketplace_tools/googleanalytics/config.yaml", "r") as file:
```
Comment (TransformerOptimus) [inline googleanalytics/userReport.py]: this is not correct to handle the config. Come up with better approach

Comment (TransformerOptimus) [reply googleanalytics/userReport.py]: use relative path.

File: googleanalytics/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+pydantic==1.10.8
+PyYAML==6.0
+superagi-tools==1.0.6
+google-analytics-data
```
Comment (TransformerOptimus) [inline googleanalytics/requirements.txt]: which version u are using for google-analytics-data

File: googleanalytics/userReport.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            dimensions=dimensions_list,
+            metrics=metrics_list,
+            date_ranges=[DateRange(start_date=start, end_date=end)],
+            limit=100000,
+            offset=0,
+        )
+
+    def return_dim_metrics(self):
+        dim_metrics = []
+        # try:
```
Comment (luciferlinx101) [inline googleanalytics/userReport.py]: Remove comments

File: googleanalytics/userReport.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            for dimension_value in row.dimension_values:
+                report += dimension_value.value + " "
+            for metric_value in row.metric_values:
+                report += metric_value.value + " "
+            report += '\n'
+        return report
+
+    def _set_google_credentials(self, credentials: str):
+        writable_credentials = json.loads(json.loads(credentials))
+        with open("sample.json", "w") as outfile:
```
Comment (TransformerOptimus) [inline googleanalytics/userReport.py]: rename it to credentials.json

File: googleanalytics/googleAnalyticsReportTool.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    def _generate_filename(self, dimensions, metrics, list_of_names):
+        filename = dimensions[0] + metrics[0]
+        if filename in list_of_names:
+            filename = filename + "New"
+        filename = filename + ".txt"
+        return filename
+
+    def _create_run_report_request(self, property_id, dimensions, metrics, start, end):
+        metrics_list = [Metric(name=x) for x in metrics]
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: Replace x with metric.

File: googleanalytics/googleAnalyticsReportTool.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+from google.analytics.data_v1beta import BetaAnalyticsDataClient
+from google.analytics.data_v1beta.types import (
+    DateRange,
+    Dimension,
+    Metric,
+    RunReportRequest,
+)
+
+class GoogleAnalyticsReportToolInput(BaseModel):
+    start: str = Field(..., description="The starting date of the query, in YYYY-MM-DD format")
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: start->start_date
end->end_date
use appropriate variable name for is_single_file as discussed.

File: googleanalytics/googleAnalyticsReportTool.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+class GoogleAnalyticsReportTool(BaseTool):
+    """
+    Analytics Report Tool
+    """
+    name: str = "Analytics Report Tool"
+    args_schema: Type[BaseModel] = GoogleAnalyticsReportToolInput
+    description: str = "Return a google analytics report for the information the user requires"
+    resource_manager: Optional[FileManager] = None
+
+    def _execute(self, start: str, end: str, is_single_file: bool):
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: change variable names accordingly.

File: googleanalytics/googleAnalyticsReportTool.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        property_id = int(self.get_tool_config("PROPERTY_ID"))
+        google_credentials = self.get_tool_config("GOOGLE_CREDENTIALS_FILE")
+
+        self._set_google_credentials(google_credentials)
+
+        client = BetaAnalyticsDataClient()
+        dim_metrics = self.return_dim_metrics()
+        list_of_names = ["report.txt"] if is_single_file else []
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: filenames

File: googleanalytics/googleAnalyticsReportTool.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        for dimensions, metrics in dim_metrics:
+            if not is_single_file:
+                filename = self._generate_filename(dimensions, metrics, list_of_names)
+                list_of_names.append(filename)
+
+            request = self._create_run_report_request(property_id, dimensions, metrics, start, end)
+            response = client.run_report(request)
+            report = self._generate_report(response)
+
+            if not is_single_file:
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: Don't use redundant if statements. Club both of these into one.

File: googleanalytics/googleAnalyticsReportTool.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        list_of_names = ["report.txt"] if is_single_file else []
+
+        for dimensions, metrics in dim_metrics:
+            if not is_single_file:
+                filename = self._generate_filename(dimensions, metrics, list_of_names)
+                list_of_names.append(filename)
+
+            request = self._create_run_report_request(property_id, dimensions, metrics, start, end)
+            response = client.run_report(request)
+            report = self._generate_report(response)
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: report +=

File: googleanalytics/googleAnalyticsReportTool.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            if not is_single_file:
+                self._write_to_file(filename, report)
+                report = ""
+
+        if is_single_file:
+            self._write_to_file(list_of_names[0], report)
+
+        os.remove("credentials.json")
+
+        response_str = " ".join(list_of_names) + "."
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: Do this in one line

File: googleanalytics/googleAnalyticsReportTool.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        dim_metrics = self.return_dim_metrics()
+        list_of_names = ["report.txt"] if is_single_file else []
+
+        for dimensions, metrics in dim_metrics:
+            if not is_single_file:
+                filename = self._generate_filename(dimensions, metrics, list_of_names)
+                list_of_names.append(filename)
+
+            request = self._create_run_report_request(property_id, dimensions, metrics, start, end)
+            response = client.run_report(request)
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: name it as google_response

File: googleanalytics/googleAnalyticsReportTool.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+        if is_single_file:
+            self._write_to_file(list_of_names[0], report)
+
+        os.remove("credentials.json")
+
+        response_str = " ".join(list_of_names) + "."
+        return "Successfully wrote " + response_str
+
+    def _generate_report(self, response):
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: change response as discussed.

File: googleanalytics/googleAnalyticsReportTool.py (hunk 14 above)
Comment (jagtarcontlo) [reply googleanalytics/googleAnalyticsReportTool.py]: response = "Successfully wrote " + " ".join(list_of_names) + "."

File: googleanalytics/googleAnalyticsReportTool.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+            property=f"properties/{property_id}",
+            dimensions=dimensions_list,
+            metrics=metrics_list,
+            date_ranges=[DateRange(start_date=start, end_date=end)],
+            limit=100000,
+            offset=0,
+        )
+
+    def return_dim_metrics(self):
+        dim_metrics = []
```
Comment (jagtarcontlo) [inline googleanalytics/googleAnalyticsReportTool.py]: change get_dimensions_and_metrics
dimensions_and_metrics = []

File: googleanalytics/googleAnalyticsReportTool.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+    def _execute(self, start_date_of_query: str, end_date_of_query: str, store_report_in_single_file: bool):
+        property_id = int(self.get_tool_config("PROPERTY_ID"))
+        google_credentials = self.get_tool_config("GOOGLE_CREDENTIALS_FILE")
+
+        self._set_google_credentials(google_credentials)
+
+        client = BetaAnalyticsDataClient()
+        dimensions_and_metrics = self.get_dimensions_and_metrics()
+        filenames = []
```
Comment (CaptainLevi0007) [inline googleanalytics/googleAnalyticsReportTool.py:39]: only declare this if store_report_in_single_file is false

Comment (lalitlj) [inline googleanalytics/googleAnalyticsReportTool.py]: Checked the code as external agent, also changed a bit of logic to make returning response easier, committing.

**Review by jagtarcontlo — APPROVED** [review]

File: googleanalytics/google_analytics_report_tool.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+                self.resource_manager.write_file(filename, report)
+                report = ""
+
+        if store_report_in_single_file:
+            self.resource_manager.write_file(filenames[0], report)
+
+        os.remove("credentials.json")
+
+        response = "Successfully wrote " + ", ".join(filenames) + "."
+        return response
```
Comment (luciferlinx101) [inline googleanalytics/google_analytics_report_tool.py]: Remove the temporary variable response and directly return

File: googleanalytics/google_analytics_report_tool.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    def _set_google_credentials(self, credentials: str):
+        writable_credentials = json.loads(json.loads(credentials))
+        with open("credentials.json", "w") as outfile:
+            json.dump(writable_credentials, outfile)
+        os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = "credentials.json"
+
+    def _generate_filename(self, dimensions, metrics, filenames):
+        filename = dimensions[0] + metrics[0]
+        if filename in filenames:
+            filename = filename + "New"
```
Comment (luciferlinx101) [inline googleanalytics/google_analytics_report_tool.py]: can we do it _new?

File: googleanalytics/google_analytics_report_tool.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+        client = BetaAnalyticsDataClient()
+        dimensions_and_metrics = self.get_dimensions_and_metrics()
+
+        filenames = ["report.txt"] if store_report_in_single_file else []
+        report=""
+
+        for dimensions, metrics in dimensions_and_metrics:
+            request = self._create_run_report_request(property_id, dimensions, metrics, start_date_of_query, end_date_of_query)
+            google_response = client.run_report(request)
```
Comment (luciferlinx101) [inline googleanalytics/google_analytics_report_tool.py:45]: Remove   client = BetaAnalyticsDataClient() and directly do BetaAnalyticsDataClient().run_report(request)

Comment (luciferlinx101) [reply googleanalytics/google_analytics_report_tool.py:45]: Temporary variables should be avoided it makes functions long and complex

**Pull request merged**